### PR TITLE
Add banner to tell the user the webview needs to be updated

### PR DESCRIPTION
**Pull request by tevincent**
Click on later:

https://github.com/Infomaniak/android-kMail/assets/149579879/10abd522-d19c-4887-b3f0-217b2e69c4de

Click on update:

https://github.com/Infomaniak/android-kMail/assets/149579879/ea9cd34d-ee32-4419-9c00-80a9a48ad515

### CODE REVIEW

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

File: app/src/main/res/layout/fragment_thread_list.xml [hunk 1]
```diff
@@ -249,7 +297,7 @@
                                 android:paddingBottom="@dimen/recyclerViewPaddingBottom"
                                 app:behind_swiped_item_icon_margin="@dimen/marginStandard"
                                 app:layout_constraintBottom_toBottomOf="parent"
-                                app:layout_constraintTop_toBottomOf="@id/installUpdateDivider"
+                                app:layout_constraintTop_toBottomOf="@id/webviewVersionDivider"
```
Comment (KevinBoulongne) [inline app/src/main/res/layout/fragment_thread_list.xml:300]: Maybe use `@id/webviewVersionGroup` instead ?

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt [hunk 2]
```diff
@@ -174,6 +176,18 @@ class MainViewModel @Inject constructor(
...
+            WebViewCompat.getCurrentWebViewPackage(appContext)?.versionName?.let { versionName ->
+                val majorVersion = versionName.substring(0, versionName.indexOfFirst { it == '.' }).toInt()
+                // Current major version of Android WebView System
+                majorVersion < WEBVIEW_MIN_VERSION
+            } ?: false
+        } else {
+            false
+        }
+    }
+
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt:190]: Does it really need to be inside the MainViewModel? ThreadListViewModel seems more fitting

File: app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt [hunk 3]
```diff
@@ -174,6 +176,18 @@ class MainViewModel @Inject constructor(
 
     private var currentThreadsLiveJob: Job? = null
 
+    fun checkWebViewVersion(showWebViewOutdated: Boolean) {
+        webViewOutdated.value = if (showWebViewOutdated) {
+            WebViewCompat.getCurrentWebViewPackage(appContext)?.versionName?.let { versionName ->
+                val majorVersion = versionName.substring(0, versionName.indexOfFirst { it == '.' }).toInt()
+                // Current major version of Android WebView System
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt:183]: What does current mean in your comment?

File: app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt [hunk 4]
```diff
@@ -109,6 +110,7 @@ class MainViewModel @Inject constructor(
     val newFolderResultTrigger = MutableLiveData<Unit>()
     val reportPhishingTrigger = SingleLiveEvent<Unit>()
     val canInstallUpdate = MutableLiveData(false)
+    val webViewOutdated = MutableLiveData(false)
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt:113]: maybe we want to call it "isWebViewOutdated"?

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt [hunk 5]
```diff
@@ -118,48 +124,59 @@ class ThreadListFragment : TwoPaneFragment(), SwipeRefreshLayout.OnRefreshListen
...
+            observeFlushFolderTrigger()
+            observeUpdateInstall()
+            observeWebViewOutdated()
+            observeLoadMoreTriggers()
+
+            viewLifecycleOwner.lifecycleScope.launch {
+                repeatOnLifecycle(State.STARTED) {
+                    mainViewModel.checkWebViewVersion(localSettings.canShowWebViewOutdated)
+                }
+            }
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt:177]: Use the Fragment's `onStart()`

File: app/src/main/java/com/infomaniak/mail/data/LocalSettings.kt [hunk 6]
```diff
@@ -66,6 +66,7 @@ class LocalSettings private constructor(context: Context) : SharedValues {
     var isMatomoTrackingEnabled by sharedValue("isMatomoTrackingEnabledKey", true)
     var autoAdvanceMode by sharedValue("autoAdvanceModeKey", AutoAdvanceMode.THREADS_LIST)
     var autoAdvanceNaturalThread by sharedValue("autoAdvanceNaturalThreadKey", AutoAdvanceMode.FOLLOWING_THREAD)
+    var canShowWebViewOutdated by sharedValue("canShowWebViewOutdated", true)
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/data/LocalSettings.kt:69]: Would "hasWebViewOutdatedBeenDismissed" or something in this direction be more appropriated?

**Review by LunarX — CHANGES_REQUESTED** [review]

File: app/src/main/res/layout/fragment_thread_list.xml (hunk 1 above)
Comment (tevincent) [reply app/src/main/res/layout/fragment_thread_list.xml:300]: It's not working for group.

File: app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt (hunk 3 above)
Comment (tevincent) [reply app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt:183]: Current mean the latest version available on the store.

File: app/src/main/java/com/infomaniak/mail/data/LocalSettings.kt (hunk 6 above)
Comment (tevincent) [reply app/src/main/java/com/infomaniak/mail/data/LocalSettings.kt:69]: Above, we also have `canShow...` that's why I prefered to keep the same logic.

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/data/LocalSettings.kt (hunk 6 above)
Comment (LunarX) [reply app/src/main/java/com/infomaniak/mail/data/LocalSettings.kt]: Maybe `show...` then

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/data/LocalSettings.kt (hunk 6 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/data/LocalSettings.kt]: Boolean should start with can, should, has, etc…

Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/data/LocalSettings.kt]: I've unresolved this conversation.
A boolean should start with is, can, should, has, etc…

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt [hunk 7]
```diff
@@ -17,28 +17,33 @@
...
 import kotlinx.coroutines.*
 import javax.inject.Inject
 
 @HiltViewModel
 class ThreadListViewModel @Inject constructor(
+    application: Application,
     private val searchUtils: SearchUtils,
     @IoDispatcher private val ioDispatcher: CoroutineDispatcher,
-) : ViewModel() {
+) : AndroidViewModel(application) {
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt:39]: No need to turn the view model into an AndroidViewModel, you can directly inject the app context with Hilt like in other view models

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt [hunk 8]
```diff
@@ -58,4 +63,20 @@ class ThreadListViewModel @Inject constructor(
         // Delete Search data in case they couldn't be deleted at the end of the previous Search.
         searchUtils.deleteRealmSearchData()
     }
+
+    fun checkWebViewVersion(showWebViewOutdated: Boolean) {
+        isWebViewOutdated.value = if (showWebViewOutdated) {
+            WebViewCompat.getCurrentWebViewPackage(appContext)?.versionName?.let { versionName ->
+                val majorVersion = versionName.substring(0, versionName.indexOfFirst { it == '.' }).toInt()
+                // First know version of Android WebView System without bug
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt:71]:
```suggestion
                // First known version of Android WebView System without bug
```

File: app/src/main/java/com/infomaniak/mail/views/itemViews/BannerWithActionView.kt [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+                descriptionIcon.isVisible = true
+                descriptionIcon.setCompoundDrawablesWithIntrinsicBounds(it, null, null, null)
+            }
+
+            actionButton.icon = attributes.getDrawable(R.styleable.BannerWithActionView_buttonIcon)
+        }
+
+        description = attributes.getString(R.styleable.BannerWithActionView_description) ?: ""
+        actionButtonText = attributes.getString(R.styleable.BannerWithActionView_buttonText) ?: ""
+    }
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/views/itemViews/BannerWithActionView.kt:96]: If you want to move the init of the view in its own method rather than doing it directly inside `init`. Place it right after `init` so we can easily find it at the top of the class like other custom views

File: app/src/main/res/layout/fragment_thread_list.xml [hunk 10]
```diff
@@ -277,6 +250,18 @@
...
+                    android:layout_width="match_parent"
+                    android:layout_height="match_parent"
+                    android:visibility="gone"
+                    app:layout_anchor="@+id/swipeRefreshLayout"
+                    app:layout_anchorGravity="center"
+                    tools:description="@string/emptyStateInboxDescription"
+                    tools:icon="@drawable/ic_empty_state_folder"
+                    tools:title="@string/emptyStateInboxTitle"
+                    tools:visibility="visible" />
+
```
Comment (LunarX) [inline app/src/main/res/layout/fragment_thread_list.xml:264]: Why did you need to move the EmptyStateView outside the FrameLayout?

File: app/src/main/res/layout/item_banner_with_action_view.xml [hunk 11]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="utf-8"?><!--
+  ~ Infomaniak Mail - Android
+  ~ Copyright (C) 2023-2024 Infomaniak Network SA
```
Comment (LunarX) [inline app/src/main/res/layout/item_banner_with_action_view.xml:3]: You can fix the copyright that came from the copy/paste by only writing 2024

**Review by LunarX — CHANGES_REQUESTED** [review]

File: app/src/main/res/layout/fragment_thread_list.xml (hunk 10 above)
Comment (tevincent) [reply app/src/main/res/layout/fragment_thread_list.xml:264]: Not sure how it happened. Thanks.

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt [hunk 12]
```diff
@@ -58,4 +63,20 @@ class ThreadListViewModel @Inject constructor(
         // Delete Search data in case they couldn't be deleted at the end of the previous Search.
         searchUtils.deleteRealmSearchData()
     }
+
+    fun checkWebViewVersion(showWebViewOutdated: Boolean) {
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt:67]: Boolean name should start with is, can, should, has, etc…

File: app/src/main/res/values-fr/strings.xml [hunk 13]
```diff
@@ -227,6 +227,8 @@
     <string name="createFolderNoParent">Aucun</string>
     <string name="createFolderParent">Dossier parent</string>
     <string name="dateTitle">Date :</string>
+    <string name="displayMailIssueDescription">Pour corriger les problèmes d’affichages, veuillez mettre à jour l’application Android System Webview.</string>
```
Comment (KevinBoulongne) [inline app/src/main/res/values-fr/strings.xml:230]:
```suggestion
    <string name="displayMailIssueDescription">Pour corriger les problèmes d’affichage, veuillez mettre à jour l’application Android System Webview.</string>
```

Comment (KevinBoulongne) [reply app/src/main/res/values-fr/strings.xml:230]: Update the string on Loco.

File: app/src/main/java/com/infomaniak/mail/views/itemViews/BannerWithActionView.kt [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+            connect(R.id.actionButton, ConstraintSet.TOP, ConstraintSet.PARENT_ID, ConstraintSet.TOP)
+            connect(R.id.actionButton, ConstraintSet.BOTTOM, ConstraintSet.PARENT_ID, ConstraintSet.BOTTOM)
+        }
+    }
+
+    init {
+        attrs?.getAttributes(context, R.styleable.BannerWithActionView) { initView(attributes = this) }
+    }
+
+
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/views/itemViews/BannerWithActionView.kt:73]:
```suggestion
```

File: app/src/main/java/com/infomaniak/mail/views/itemViews/BannerWithActionView.kt [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.infomaniak.mail.views.itemViews
+
+import android.content.Context
+import android.content.res.TypedArray
+import android.graphics.drawable.Drawable
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/views/itemViews/BannerWithActionView.kt:22]:
```suggestion
```

File: app/src/main/res/values/attrs.xml [hunk 16]
```diff
@@ -98,6 +98,14 @@
         <attr name="ripple" format="boolean" />
     </declare-styleable>
 
+    <declare-styleable name="BannerWithActionView">
+        <attr name="description"/>
```
Comment (KevinBoulongne) [inline app/src/main/res/values/attrs.xml:102]:
```suggestion
        <attr name="description" />
```

File: app/src/main/java/com/infomaniak/mail/views/itemViews/BannerWithActionView.kt [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+
+    init {
+        attrs?.getAttributes(context, R.styleable.BannerWithActionView) { initView(attributes = this) }
+    }
+
+
+    private fun initView(attributes: TypedArray) {
+        with(binding) {
+            if (attributes.getBoolean(R.styleable.BannerWithActionView_isButtonAlignedWithDescription, false)) {
+                binding.description.setMarginsRelative(end = resources.getDimensionPixelSize(R.dimen.alternativeMargin))
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/views/itemViews/BannerWithActionView.kt:77]:
```suggestion
                description.setMarginsRelative(end = resources.getDimensionPixelSize(R.dimen.alternativeMargin))
```

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt [hunk 18]
```diff
@@ -19,6 +19,7 @@ package com.infomaniak.mail.ui
 
 import android.app.Application
 import androidx.lifecycle.*
+import androidx.webkit.WebViewCompat
```
[the remark this answers is not in the document]
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt]: "Reformat code" on whole project again 😁

**Review by KevinBoulongne — APPROVED** [review]

File: app/src/main/res/values/attrs.xml [hunk 19]
```diff
@@ -98,6 +98,14 @@
         <attr name="ripple" format="boolean" />
     </declare-styleable>
 
+    <declare-styleable name="ItemBannerWithButtonActionView">
+        <attr name="description"/>
+        <attr name="descriptionIcon" format="reference" />
+        <attr name="buttonText" format="string" />
+        <attr name="buttonIcon" format="reference" />
+        <attr name="isButtonAlignWithDescription" format="boolean" />
```
[the remark this answers is not in the document]
Comment (LunarX) [reply app/src/main/res/values/attrs.xml]: And rename its usages as well. You still use the old naming inside `item_banner_with_action_view`

File: app/src/main/res/layout/fragment_thread_list.xml [hunk 20]
```diff
@@ -260,6 +241,8 @@
                             android:layout_width="match_parent"
                             android:layout_height="match_parent"
                             android:visibility="gone"
+                            app:layout_anchor="@+id/swipeRefreshLayout"
+                            app:layout_anchorGravity="center"
```
Comment (LunarX) [inline app/src/main/res/layout/fragment_thread_list.xml:245]: I don't think you need to add these lines

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt [hunk 21]
```diff
@@ -58,4 +63,20 @@ class ThreadListViewModel @Inject constructor(
...
+    fun checkWebViewVersion(canShowWebViewOutdated: Boolean) {
+        isWebViewOutdated.value = if (canShowWebViewOutdated) {
+            WebViewCompat.getCurrentWebViewPackage(appContext)?.versionName?.let { versionName ->
+                val majorVersion = versionName.substring(0, versionName.indexOfFirst { it == '.' }).toInt()
+                // First known version of Android WebView System without bug
+                majorVersion < WEBVIEW_MIN_VERSION
+            } ?: false
+        } else {
+            false
+        }
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListViewModel.kt:76]: Shouldn't we handle the case where it can't be parsed as an Int so the app doesn't crash?

File: app/src/main/java/com/infomaniak/mail/views/itemViews/BannerWithActionView.kt [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+        set(value) {
+            binding.actionButton.text = value
+        }
+
+    private val baseConstraints by lazy {
+        ConstraintSet().apply {
+            clone(binding.root)
+        }
+    }
+    private val buttonAlignConstraints by lazy {
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/views/itemViews/BannerWithActionView.kt:57]:
```suggestion
    }

    private val buttonAlignConstraints by lazy {
```

File: app/src/main/java/com/infomaniak/mail/views/itemViews/BannerWithActionView.kt [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+    defStyleAttr: Int = 0,
+) : ConstraintLayout(context, attrs, defStyleAttr) {
+
+    private val binding by lazy { ViewBannerWithActionBinding.inflate(LayoutInflater.from(context), this, true) }
+
+    var description: CharSequence?
+        get() = binding.description.text
+        set(value) {
+            binding.description.text = value
+        }
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/views/itemViews/BannerWithActionView.kt:44]: Use delegation if the signatures are the exact same ones

```suggestion
    var description: CharSequence? by binding.actionButton::text
```

File: app/src/main/java/com/infomaniak/mail/views/itemViews/BannerWithActionView.kt [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+
+    var actionButtonText: CharSequence?
+        get() = binding.actionButton.text
+        set(value) {
+            binding.actionButton.text = value
+        }
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/views/itemViews/BannerWithActionView.kt:50]: Same here

```suggestion
    var actionButtonText: CharSequence? by binding.actionButton::text
```

File: app/src/main/res/layout/fragment_thread_list.xml
Comment (LunarX) [inline app/src/main/res/layout/fragment_thread_list.xml:1]: Do we still need a ConstrainLayout to contain the two banners and the DragDropSwipeRecyclerView?

File: app/src/main/res/layout/item_banner_with_action_view.xml [hunk 25]
```diff
@@ -0,0 +1,26 @@
...
+  -->
+<com.infomaniak.mail.views.itemViews.BannerWithActionView xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    app:buttonIcon="@drawable/ic_bin"
+    tools:buttonText="@string/threadListEmptySpamButton"
+    tools:description="@string/emptyStateTrashDescription"
+    tools:isButtonAlignWithDescription="false" />
```
Comment (LunarX) [inline app/src/main/res/layout/item_banner_with_action_view.xml:26]: The property hasn't been renamed here

**Review by LunarX — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/views/itemViews/BannerWithActionView.kt (hunk 22 above)
Comment (tevincent) [reply app/src/main/java/com/infomaniak/mail/views/itemViews/BannerWithActionView.kt:57]: I prefer to keep them without a new line because these 2 variable are linked.

File: app/src/main/res/layout/fragment_thread_list.xml
Comment (tevincent) [reply app/src/main/res/layout/fragment_thread_list.xml:1]: I can remove the `FrameLayout` and put everything in the `ConstraintLayout`.

Comment (LunarX) [reply app/src/main/res/layout/fragment_thread_list.xml:1]: No, the `FrameLayout` was good as it was. We should keep using it to show that we either display the empty state or the "filled" layout. The only thing you could have replaced was the `ConstraintLayout` with a `LinearLayout` since every three items where simply stacked on top of one another, no need to bother with constraints in this case

**Review by LunarX — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/views/itemViews/BannerWithActionView.kt (hunk 23 above)
Comment (tevincent) [reply app/src/main/java/com/infomaniak/mail/views/itemViews/BannerWithActionView.kt:44]: `var description: CharSequence? by binding.description::text` and not `actionButton`. Will do the modification

Comment (LunarX) [reply app/src/main/java/com/infomaniak/mail/views/itemViews/BannerWithActionView.kt]: Yeah true nice catch
